### ISSUE TITLE
linear function and the different convs functions retain unnecessary buffers

### ISSUE THREAD

**Issue by rotabulo** [pytorch/pytorch#42371]
## 🐛 Bug

The function `torch.nn.functional` but also the different `torch.nn.functional.convXd` keep allocated the input also when the weight does not require the gradient.

## To Reproduce

The following code shows the issue by comparing `torch.nn.functional` against a custom implementation of it.

```
import torch

from torch.nn.functional import linear


# Creating the customized function
class MyLinear(torch.autograd.Function):
    @staticmethod
    def forward(ctx, i, w):

        if ctx.needs_input_grad[1]:
            ctx.save_for_backward(w, i)
        else:
            ctx.save_for_backward(w)
        return i.mm(w.t())

    @staticmethod
    def backward(ctx, grad_output):
        saved = ctx.saved_tensors
        return grad_output.mm(saved[0]), grad_output.t().mm(saved[1]) if ctx.needs_input_grad[1] else None

# Input requiring gradient
x = torch.rand(1000, 10).cuda().requires_grad_()

# Weights NOT requiring gradient
weight = torch.rand(20, 10).cuda()


# This function performs some operations using the function f and keeps track of the
# allocated gpu memory within the function.
def test(f, x, weight):
    s = torch.cuda.memory_allocated()
    x = x + 2
    x = f(x, weight)
    print(f"Allocated memory: {torch.cuda.memory_allocated() - s}")
    return x


# I report the amount of allocated memory by torch.nn.functional
print("Test Pytorch linear")
y1 = test(linear, x, weight)

# I report the amount of allocated memory by the customized implementation
print("\nTest MyLinear")
y2 = test(MyLinear.apply, x, weight)

# Check that the obtained outputs agree
print(f"\nSame output? {y1.allclose(y2)}")

y1.norm().backward()
z1 = x.grad.clone()

x.grad.zero_()
y2.norm().backward()
z2 = x.grad.clone()

# Check that the gradient wrt input agrees.
print(f"Same gradient? {z1.allclose(z2)}")
```

The output I get:
```
Test Pytorch linear
Allocated memory: 120832

Test MyLinear
Allocated memory: 80384

Same output? True
Same gradient? True
```

## Expected behavior

`torch.nn.autograd.linear` should return the same amount of memory allocated by the custom implementation.
Specifically, when the weight does not require the gradient, the input should not be stored for later use.

NOTE: this applies not just to `linear` but also to the convolution operator of Pytorch.

## Environment

```
Collecting environment information...
PyTorch version: 1.6.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 20.04.1 LTS
GCC version: (Ubuntu 8.4.0-3ubuntu2) 8.4.0
CMake version: version 3.16.3

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 440.100
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.3
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.5.1

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen @mruberry

**Comment by ngimel** [comment]
This is because of how `mm` derivative is defined https://github.com/pytorch/pytorch/blob/bfa94487b968ccb570ef8cd9547029b967e76ed0/tools/autograd/derivatives.yaml#L700 - it references `self` tensor instead of just its sizes and strides. It should not be necessary - the backward function uses only sizes, strides and is_sparse from `self` argument, so that's what backward formula should be capturing 
https://github.com/pytorch/pytorch/blob/bfa94487b968ccb570ef8cd9547029b967e76ed0/tools/autograd/templates/Functions.cpp#L613-L617

**Comment by gchanan** [comment]
probably the codegen isn't smart enough to save is_sparse, so we'd have to implement that.

**Comment by mruberry** [comment]
Triage review believes this may be happening in more places, too.

**Comment by albanD** [comment]
>  it references self tensor instead of just its sizes and strides. It should not be necessary 

Note that if we change the function to use `self.size()`, then the codegen will do the right thing and stop saving `self` and will only keep a copy of the size (and stride) around.

For convolution, it is trickier because we have a single formula and we don't specify in the formula what is needed or not depending on which input requires grad.
We may want to add that.

**Comment by kurtamohler** [comment]
I tried changing `mm_mat1_backward` to take the sizes, strides, and is_sparse rather than the entire `mat1` argument. But this didn't seem to change the amount of memory used. When I run the repro script from above:

```
Test Pytorch linear
Allocated memory: 120832

Test MyLinear
Allocated memory: 80384

Same output? True
Same gradient? True
```

Perhaps I misunderstand the proposed solution.

**Comment by kurtamohler** [comment]
The generated `AddmmBackward` and `MmBackward` Node objects still hold onto the `mat2` Tensor and reference it in their `apply()` functions in `torch/csrc/autograd/generated/Functions.cpp`.

I can see how we could create custom backward nodes, which would also mean we'd have to override the generated forward functions in `torch/csrc/autograd/generated/VariableTypeEverything.cpp`. Is there an automatic way to do this, rather than manually writing the custom forward/backward functions?

**Comment by ngimel** [comment]
But mat1_backward legitimately needs mat2 (not just its sizes) and vice versa, mat2_backward needs m1.

**Comment by kurtamohler** [comment]
Sorry, I meant `mat1` above. I've edited that typo.

**Comment by kurtamohler** [comment]
That's a good point though. Since `mm_mat1_backward` requires `mat1`, `mm_mat2_backward` requires `mat2`, and since both of these backward functions are being called from the same `AddmmBackward::apply()` function, it doesn't seem very simple to tell codegen to avoid keeping a reference to either `mat1` or `mat2` if one of their backward functions won't be called.

**Comment by albanD** [comment]
When multiple formulas are provided, you could update the codegen to only save Tensor if the corresponding input requires gradients. And only unpack it in the backward if the corresponding backward call needs to be done.
But this might be tricky codegen to write in general (which is most likely why we don't do it right now).

**Comment by kurtamohler** [comment]
> you could update the codegen to only save Tensor if the corresponding input requires gradients.

@albanD yes, that makes sense. In order to accomplish that in this case, we would first need to avoid any `mat1` tensor references for the generated code in `Functions.cpp` that calls `mm_mat1_backward`. My in-progress PR #45777 does this, would you mind taking a look if you have some time and let me know your thoughts?

**Comment by kurtamohler** [comment]
I just realized that my PR #45777 is not just a first step to a solution--it's the entire solution. After that change, the mm backward function can automatically avoid retaining the unnecessary tensor data. I had assumed codegen didn't have this ability yet, but evidently not. Running the repro script in the issue description, I get this now:
```
Test Pytorch linear
Allocated memory: 80384

Test MyLinear
Allocated memory: 80384

Same output? True
Same gradient? True
```

I wonder if there are any other operations we could optimize as well. Either way, we can close this issue now.
